From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Tue, 15 Jul 2025 18:56:59 +0200
Subject: apt: Handle gstreamer ()(64bit) suffix on any architecture

The ()(64bit) suffix is an RPM artifact, where it is used to distinguish
between the primary and secondary architecture in a 2-arch system.
It doesn't translate precisely to DPKG, where multi-arch is not limited
to one secondary architecture. Let's use it the same way RPM would and
only match the native arch.

Bug-Ubuntu: https://launchpad.net/bugs/2060730
Origin: https://github.com/PackageKit/PackageKit/pull/875
---
 backends/apt/apt-job.cpp     |  4 +++-
 backends/apt/gst-matcher.cpp | 25 +++++++++++++++----------
 backends/apt/gst-matcher.h   |  4 ++--
 3 files changed, 20 insertions(+), 13 deletions(-)

diff --git a/backends/apt/apt-job.cpp b/backends/apt/apt-job.cpp
index 1a89dee..2027ca4 100644
--- a/backends/apt/apt-job.cpp
+++ b/backends/apt/apt-job.cpp
@@ -629,13 +629,15 @@ void AptJob::providesCodec(PkgList &output, gchar **values)
         }
 
         arch = string(ver.Arch());
+        bool native = arch == "all" ||
+                      arch == m_cache->GetPkgCache()->NativeArch();
 
         pkgCache::VerFileIterator vf = ver.FileList();
         pkgRecords::Parser &rec = m_cache->GetPkgRecords()->Lookup(vf);
         const char *start, *stop;
         rec.GetRec(start, stop);
         string record(start, stop - start);
-        if (matcher.matches(record, arch)) {
+        if (matcher.matches(record, native)) {
             output.append(ver);
         }
     }
diff --git a/backends/apt/gst-matcher.cpp b/backends/apt/gst-matcher.cpp
index 75a342b..dd5aac3 100644
--- a/backends/apt/gst-matcher.cpp
+++ b/backends/apt/gst-matcher.cpp
@@ -52,7 +52,8 @@ GstMatcher::GstMatcher(gchar **values)
         regmatch_t matches[6];
         if (regexec(&pkre, value, 6, matches, 0) != REG_NOMATCH) {
             Match values;
-            string version, type, data, opt, arch;
+            string version, type, data, opt;
+            bool native = false;
 
             // Appends the version "0.10"
             version = "\nGstreamer-Version: ";
@@ -71,12 +72,16 @@ GstMatcher::GstMatcher(gchar **values)
                 if (!opt.empty()) {
                     size_t start_pos = 0;
                     // This is hardcoded in pk-gstreamer-install, so we also hardcode it here
-                    const string x86_64 = ")(64bit";
-
-                    if (ends_with(opt.c_str(), x86_64.c_str())) {
-                            // We hardcode 64bit -> amd64 here
-                            arch = "amd64";
-                            opt.erase(opt.end() - x86_64.length(), opt.end());
+                    const string arch_64 = ")(64bit";
+                    if (ends_with(opt.c_str(), arch_64.c_str())) {
+                        /* The ()(64bit) suffix is an RPM artifact, where it is used to distinguish
+                         * between the primary and secondary architecture in a 2-arch system.
+                         * It doesn't translate precisely to DPKG, where multi-arch is not limited
+                         * to one secondary architecture. Let's use it the same way RPM would and
+                         * only match the native arch.
+                         */
+                        native = true;
+                        opt.erase(opt.end() - arch_64.length(), opt.end());
                     }
 
                     // Replace all ")(" with "," - convert from input to serialized caps format
@@ -122,7 +127,7 @@ GstMatcher::GstMatcher(gchar **values)
             values.data    = data;
             values.opt     = opt;
             values.caps    = caps;
-            values.arch    = arch;
+            values.native  = native;
 
             m_matches.push_back(values);
         } else {
@@ -139,13 +144,13 @@ GstMatcher::~GstMatcher()
     }
 }
 
-bool GstMatcher::matches(string record, string arch)
+bool GstMatcher::matches(string record, bool native)
 {
     for (const Match &match : m_matches) {
         // Tries to find "Gstreamer-version: xxx"
         if (record.find(match.version) != string::npos) {
             size_t found;
-            if (!match.arch.empty() && arch != match.arch)
+            if (match.native && !native)
                     continue;
             found = record.find(match.type);
             // Tries to find the type "Gstreamer-Uri-Sinks: "
diff --git a/backends/apt/gst-matcher.h b/backends/apt/gst-matcher.h
index 9d3615c..7c696bb 100644
--- a/backends/apt/gst-matcher.h
+++ b/backends/apt/gst-matcher.h
@@ -34,7 +34,7 @@ typedef struct {
     string   data;
     string   opt;
     void    *caps;
-    string   arch;
+    bool     native;
 } Match;
 
 class GstMatcher
@@ -43,7 +43,7 @@ public:
     GstMatcher(gchar **values);
     ~GstMatcher();
 
-    bool matches(string record, string arch);
+    bool matches(string record, bool arch);
     bool hasMatches() const;
 
 private:
